internal/checks/latency.go

```diff
@@ -0,0 +1,484 @@
+	}
+}
+
+// NewCircuitBreaker abstracts away the rather complex circuitbreaker.Builder API.
+// https://pkg.go.dev/github.com/failsafe-go/failsafe-go/circuitbreaker
+// https://failsafe-go.dev/circuit-breaker/
+func NewCircuitBreaker(
+	errorRate float64,
+	detectionWindow time.Duration,
+	banWindow time.Duration,
+) circuitbreaker.CircuitBreaker[any] {
+	// TODO(polsar): Check that `0.0 < errorRate <= 1.0` holds.
+	return circuitbreaker.Builder[any]().
+		HandleResult(false). // The false return value of the wrapped call will be interpreted as a failure.
+		WithFailureRateThreshold(
+			uint(math.Floor(errorRate*PercentPerFrac)), // Minimum percentage of failed requests to open the breaker.
+			MinNumRequestsForRate,
+			detectionWindow,
+		).
+		WithDelay(banWindow).
+		Build()
+}
+
+type LatencyCheck struct {
+	client               client.EthClient
+	Err                  error
+	clientGetter         client.EthClientGetter
+	metricsContainer     *metrics.Container
+	logger               *zap.Logger
+	upstreamConfig       *conf.UpstreamConfig
+	routingConfig        *conf.RoutingConfig
+	errorCircuitBreaker  ErrorCircuitBreaker
+	methodLatencyBreaker map[string]LatencyCircuitBreaker // RPC method -> LatencyCircuitBreaker
+	lock                 sync.RWMutex
+	ShouldRun            bool
+}
+
+func NewLatencyChecker(
+	upstreamConfig *conf.UpstreamConfig,
+	routingConfig *conf.RoutingConfig,
+	clientGetter client.EthClientGetter,
+	metricsContainer *metrics.Container,
+	logger *zap.Logger,
+) types.LatencyChecker {
+	c := &LatencyCheck{
+		upstreamConfig:       upstreamConfig,
+		routingConfig:        routingConfig,
+		clientGetter:         clientGetter,
+		metricsContainer:     metricsContainer,
+		logger:               logger,
+		errorCircuitBreaker:  NewErrorStats(routingConfig),
+		methodLatencyBreaker: make(map[string]LatencyCircuitBreaker),
+		ShouldRun:            routingConfig.Errors != nil || routingConfig.Latency != nil,
+	}
+
+	if err := c.InitializePassiveCheck(); err != nil {
+		logger.Error("Error initializing LatencyCheck.", zap.Any("upstreamID", c.upstreamConfig), zap.Error(err))
+	}
+
+	return c
+}
+
+func (c *LatencyCheck) InitializePassiveCheck() error {
+	// TODO(polsar): Set `c.ShouldRun` if active health checking is enabled.
+	c.logger.Debug("Initializing LatencyCheck.", zap.Any("config", c.upstreamConfig))
+
+	httpClient, err := c.clientGetter(c.upstreamConfig.HTTPURL, &c.upstreamConfig.BasicAuthConfig, &c.upstreamConfig.RequestHeadersConfig)
+	if err != nil {
+		c.Err = err
+		return c.Err
+	}
+
+	c.client = httpClient
+
+	c.runPassiveCheck()
+
+	// TODO(polsar): This check is in both PeerCheck and SyncingCheck implementations, so refactor this.
+	if isMethodNotSupportedErr(c.Err) {
+		c.logger.Debug("LatencyCheck is not supported by upstream, not running check.", zap.String("upstreamID", c.upstreamConfig.ID))
+
+		c.ShouldRun = false
+	}
+
+	return nil
+}
+
+func (c *LatencyCheck) RunPassiveCheck() {
+	if c.client == nil {
```

Comment:
I feel like we should check `c.ShouldRun` before trying to initialize the client